Add support for NXP S32 CAN driver

congnguyenhuu opened the pull request:
This patch introduces support for NXP S32 CANEXCEL (CANXL) peripheral. CAN protocol supporting:
- CAN classic
- CAN FD

Currently, remote transmission request is not supported as this feature is not available on NXP S32 CANXL HAL.

Updates in NXP HAL: https://github.com/zephyrproject-rtos/hal_nxp/pull/206

henrikbrixandersen commented:
@congnguyenhuu Thank you for your contribution. However, please split the suggested API change for the CAN RX filters into a separate PR so that it can be reviewed stand-alone (the current approach will cause all other boards to fail as they do not implement filtering of FD frames). You can put this PR as draft until a decision regarding the API change has been made and merged.

henrikbrixandersen reviewed (changes requested):
@congnguyenhuu Thank you for your contribution. However, please split the suggested API change for the CAN RX filters into a separate PR so that it can be reviewed stand-alone (the current approach will cause all other boards to fail as they do not implement filtering of FD frames). You can put this PR as draft until a decision regarding the API change has been made and merged.

congnguyenhuu committed:
boards: s32z270dc2_r52: add CAN document to list of supported features

Add CAN document to the list of supported features

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
boards: arm: s32z270dc2_r52: enable CAN support

Enable CAN instances on s32z270dc2_r52 boards

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
drivers: can: support NXP S32 CANEXCEL

This patch introduces support for NXP S32 CANEXCEL (CANXL) peripheral.

CAN protocol supporting:
- CAN classic
- CAN FD

Remote transmission request is not supported as this feature is not
available on NXP S32 CANXL HAL.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu commented:
@henrikbrixandersen Could you pls help me review.

manuargue reviewed (commented):
HAL pr was merged, please update the manifest

congnguyenhuu commented:
> HAL pr was merged, please update the manifest
I updated

dleach02 commented:
@henrikbrixandersen, can you revisit to determine of your blocking request has been resolved? The submitter is trying to get this PR merged before 3.3 feature freeze

henrikbrixandersen commented on boards/arm/s32z270dc2_r52/s32z270dc2_r52.dtsi:
```diff
@@ -89,3 +89,15 @@
+	pinctrl-0 = <&can0_default>;
+	pinctrl-names = "default";
+	clock-frequency = <80000000>;
+	bus-speed = <1000000>;
```
Can this be changed to the Zephyr "default" of 125000?

congnguyenhuu replied:
yes, this can. This is max value.

henrikbrixandersen commented on boards/arm/s32z270dc2_r52/s32z270dc2_r52.dtsi:
```diff
@@ -89,3 +89,15 @@
+	clock-frequency = <80000000>;
+	bus-speed = <1000000>;
+	sample-point = <875>;
+	sjw = <2>;
```
Can this be changed to the Zephyr "default" of 1?

congnguyenhuu replied:
yes, can

henrikbrixandersen commented on boards/arm/s32z270dc2_r52/s32z270dc2_r52.dtsi:
```diff
@@ -89,3 +89,15 @@
+	bus-speed = <1000000>;
+	sample-point = <875>;
+	sjw = <2>;
+	bus-speed-data = <8000000>;
```
Can this be changed to the Zephyr "default" of 1000000?

congnguyenhuu replied:
we can do, but this is max value

henrikbrixandersen commented on boards/arm/s32z270dc2_r52/s32z270dc2_r52.dtsi:
```diff
@@ -89,3 +89,15 @@
+	sjw = <2>;
+	bus-speed-data = <8000000>;
+	sample-point-data = <875>;
+	sjw-data = <2>;
```
Can this be changed to the Zephyr "default" of 1?

congnguyenhuu replied:
yes.

henrikbrixandersen commented on drivers/can/Kconfig.nxp_s32:
```diff
@@ -0,0 +1,36 @@
+	help
+	  Maximum number of TX descriptors.
+
+config CAN_NXP_S32_BRS
+	bool "Frame uses CAN-FD Baud Rate Switch (BRS)"
+	depends on CAN_FD_MODE
+	default y
+	help
+	  Frame uses CAN-FD Baud Rate Switch.
```
Other CAN drivers allow specifying BRS on a per-frame basis. Please provide more context in the help text why this is a compile-time setting for the S32 CANXL.

congnguyenhuu replied:
NXP S32 CANXL required for set mode CAN-FD

henrikbrixandersen commented on dts/bindings/can/nxp,s32-canxl.yaml:
```diff
@@ -0,0 +1,27 @@
+  interrupts:
+    required: true
+
+  clock-frequency:
+    type: int
+    required: true
+    description: Module clock frequency in Hz.
```
Can this frequency be determined using the clock controller driver instead? Or is that not available for the S32 (yet)?

congnguyenhuu replied:
currently, the clock controller driver isn't available for NXP S32

henrikbrixandersen commented on samples/drivers/can/babbling/boards/s32z270dc2_rtu0_r52.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SAMPLE_CAN_BABBLING_FD_MODE=y
```
Why is this needed? This changes the default behaviour of the sample for this one board? Instead of adding overlays and board config files to all the CAN samples, please enable CAN on the board level. This does not scale.

congnguyenhuu replied:
yes, I enabled CAN on board level

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+		return -EALREADY;
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_enable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to enable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
```
Is it necessary to enter freeze mode in order to enable the PHY? If so, why?

This function is supposed to start the CAN controller so that it begins sending ACKs for received frames. I am unable to tell if this is done here?

congnguyenhuu replied:
yes, I removed

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+		}
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_disable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to disable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
```
Is it necessary to enter freeze mode in order to disable the PHY? If so, why?

This function is supposed to stop the CAN controller so that it does not send ACKs for any frames on the bus. I am unable to tell if this is done here?

congnguyenhuu replied:
yes, really no necessary. I removed

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+
+	is_canfd = !!(mode & CAN_MODE_FD);
+
+	if (mode & CAN_MODE_LISTENONLY) {
+		can_nxp_s32_mode = CANEXCEL_LISTEN_ONLY_MODE;
+	} else if (mode & CAN_MODE_LOOPBACK) {
+		can_nxp_s32_mode = CANEXCEL_LOOPBACK_MODE;
+	}
```
Are these mutually exclusive? If yes, the code needs to error-out if both are requested.

congnguyenhuu replied:
yes, I added condition to check error

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+			}
+		}
+	}
+
```
You are not filling in the bus error counters in `struct can_bus_err_cnt *err_cnt`?

congnguyenhuu replied:
NXP S32 HAL not support error counters, so can't

henrikbrixandersen replied:
Ok, that's odd. I thought they were required. Please at least set the error counters to zero, then.

congnguyenhuu replied:
I updated the counters, set to 0.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+#endif
+		.idType = !!(filter->flags & CAN_FILTER_IDE) ?
+				CANEXCEL_MSG_ID_EXT : CANEXCEL_MSG_ID_STD,
+		.dataLength = 64,
```
Could we use a define here instead of a "magic number"?

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+	int alloc;
+	void *arg;
+
+	switch (eventType) {
+	case CANEXCEL_EVENT_TX_WARNING:
+		LOG_WRN("Tx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_RX_WARNING:
+		LOG_WRN("Rx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_BUSOFF:
+		LOG_ERR("Bus Off (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR:
+		LOG_ERR("Error Format Frames (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR_FD:
+		LOG_ERR("Error Data Phase (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_PASSIVE:
+		LOG_ERR("Error Passive (error 0x%x)", u32SysStatus);
+		break;
+	default:
+		break;
+	}
```
Either remove these or change them to `LOG_DBG()`. This will spam the log in perfectly valid network situations.

congnguyenhuu replied:
I changed to LOG_DBG

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+	memcpy(frame->data, msg_data.data, can_dlc_to_bytes(frame->dlc));
+
+#ifdef CONFIG_CAN_RX_TIMESTAMP
+	frame->timestamp = msg_data.timeStampH << 32 | msg_data.timeStampL;
```
`frame->timestamp` is 16 bit. Are you sure this is correct?

congnguyenhuu replied:
I didn't pay attention to the variable type, we can extra variable to 64bit ?
This device driver supports timestamp 64bit

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+		return err;
+	}
+
+	/* Enable CANXL HW */
+	IP_MC_RGM->PRST_0[0].PRST_0 &=
+		~(MC_RGM_PRST_0_PERIPH_16_RST_MASK | MC_RGM_PRST_0_PERIPH_24_RST_MASK);
```
Will this enable the CANXL controller and enable it to send ACKs? If so, this needs to be delayed until start() is called.

henrikbrixandersen replied:
If so, it needs to be enabled in listen-only mode so it does not disturb the bus with ACKs/error frames until explicitly enabled by calling `can_start()`.

henrikbrixandersen replied:
Please point me to where this is done?

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+#endif
+};
+
+#define CAN_NXP_S32_NODE(n)		DT_NODELABEL(can##n)
```
You need to use the DT_INST_ macros here, not DT_NODELABEL().

congnguyenhuu replied:
This is used like other drivers, supported by NXP S32

manuargue replied:
the HAL is index-based, hence the use of `DT_NODELABEL`. We'll update all existing NXP S32 shim drivers in the future to make use of `DT_INST` macros, for now I'd suggest to keep it as it is for consistency with the rest of the drivers we've added

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+					DT_REG_ADDR_BY_NAME(CAN_NXP_S32_NODE(n), grp_ctrl),	\
+		.base_dsc_ctrl = (CANXL_DSC_CONTROL_Type *)					\
+					DT_REG_ADDR_BY_NAME(CAN_NXP_S32_NODE(n), dsc_ctrl),	\
+		.instance = n,									\
```
Why is this needed?

congnguyenhuu replied:
This device driver needs to base on the hw peripheral instance number as required by the HAL

manuargue replied:
this will be addressed for all peripheral drivers for NXP S32, hopefully soon before v3.4 :)

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,977 @@
+		CAN_NXP_S32_TIMING_DATA_CONFIG(n)						\
+		.max_bitrate = DT_CAN_TRANSCEIVER_MAX_BITRATE(CAN_NXP_S32_NODE(n),		\
+								CAN_NXP_S32_MAX_BITRATE),	\
+		.phy = NULL,									\
```
Shouldn't this be a reference to the PHY?

congnguyenhuu replied:
yes, I updated

manuargue commented:
hi @henrikbrixandersen , could you please revisit the changes and check if all comments have been addressed? thanks

henrikbrixandersen commented on boards/arm/s32z270dc2_r52/s32z270dc2_rtu0_r52.yaml:
```diff
@@ -13,3 +13,5 @@ supported:
   - watchdog
   - netif:eth
+  - can
+  - canfd
```
You need to remove `- canfd` as this is no longer needed.

henrikbrixandersen commented on boards/arm/s32z270dc2_r52/s32z270dc2_rtu1_r52.yaml:
```diff
@@ -13,3 +13,5 @@ supported:
   - watchdog
   - netif:eth
+  - can
+  - canfd
```
Same here.

congnguyenhuu replied:
I removed it

henrikbrixandersen commented on include/zephyr/drivers/can.h:
```diff
@@ -173,7 +173,7 @@ struct can_frame {
 	 * @note @kconfig{CONFIG_CAN_RX_TIMESTAMP} must be selected for this
 	 * field to be available.
 	 */
-	uint16_t timestamp;
+	uint64_t timestamp;
```
I am still not 100% sold on this change.

We need to document that the full width of the timestamp field is not used by all drivers; some timestamps will wrap around at 32 bit. You also need to look into the format specifiers used (e.g. in the CAN shell). They likely need updating to match the 64 bit size?

Another option could be to mask the upper 32 bit of the timestamp on NXP CANXL, as this seems to be the only device supporting 64 bit timestamps for now?

henrikbrixandersen replied:
I have tried to survey the market for other CAN controllers supporting 64 bit timestamps, but came up short. Please leave out this change for now.

henrikbrixandersen replied:
I am sorry, my initial comment said 32 bit, where I meant the existing 16 bits. Please leave out any changes to the timestamp counter bit size from this PR and - if deemed needed - submit a separate PR for this after the S32 CAN driver PR goes in.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+	}
+
+	canfd = !!(mode & CAN_MODE_FD);
+	brs = canfd ? true : false;
```
Is this `brs` variable just for enabling BRS when receiving? TX frame BRS most not be enabled (according to the current Zephyr CAN API) just based on `CAN_MODE_FD`.

Since `canfd` is already a boolean, you could just assign its value directly to `brs`.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+	start_time = k_uptime_ticks();
+	config->base_sic->BCFG1 &= (~CANXL_SIC_BCFG1_ABRDIS_MASK);
+
+	if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
```
Not sure I understand this part. If the caller specifies `K_NO_WAIT` as timeout, we do not return whether the bus was recovered but always assume it was?

congnguyenhuu replied:
here after the caller specifies `K_NO_WAIT `as timeout. if state is still bus-off, continue to get state until over timeout.
it return `-EAGAIN`.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+
+	if (atomic_test_and_clear_bit(data->rx_allocs, filter_id)) {
+		if (can_nxp_s32_abort_msg(config, mb_indx)) {
+			LOG_ERR("Can't abort message !");
```
Should this error be propagated to the caller? Was the MB really freed?

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+	/* Set Rx Mb individual mask for */
+	mb_indx = ALLOC_IDX_TO_RXMB_IDX(alloc);
+	if (!!(filter->flags & CAN_FILTER_IDE)) {
+		mask =  (filter->mask & CANXL_IP_ID_EXT_MASK);
```
```suggestion
		mask = (filter->mask & CANXL_IP_ID_EXT_MASK);
```

congnguyenhuu replied:
I updated

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		return -EINVAL;
+	}
+
+	if ((frame->flags & CAN_FRAME_FDF) != 0) {
+		if (frame->dlc > CANFD_MAX_DLC) {
+			LOG_ERR("DLC of %d for CAN-FD format frame", frame->dlc);
+			return -EINVAL;
+		}
```
This could be `#ifdef`'ed out if CAN-FD mode isn't enabled in Kconfig.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		.idType = !!(frame->flags & CAN_FRAME_IDE) ?
+				CANEXCEL_MSG_ID_EXT : CANEXCEL_MSG_ID_STD,
+		.priority = 0,
+		.retransmission = 0,
```
Can this field be used to implement support for `CAN_MODE_ONE_SHOT`?

henrikbrixandersen replied:
So, does setting `.retransmission = 0` mean no retransmissions? If so, it is a violation of the Zephyr `can_send()` API when `CAN_MODE_ONE_SHOT` is not set.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		.is_polling = FALSE
+	};
+
+	LOG_DBG("%s: Sending %d bytes Tx Mb %d, "
+		"Tx Id: 0x%x, "
+		"Id type: %s %s %s %s"
+		, dev->name, data_length
+		, mb_indx, frame->id
+		, !!(frame->flags & CAN_FRAME_IDE) ?
+						"extended" : "standard"
+		, !!(frame->flags & CAN_FRAME_RTR) ? "RTR" : ""
+		, !!(frame->flags & CAN_FRAME_FDF) ? "FD frame" : ""
+		, !!(frame->flags & CAN_FRAME_BRS) ? "BRS" : "");
```
Please fix the formatting of these lines with the comma being the last character instead of the first.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		, !!(frame->flags & CAN_FRAME_BRS) ? "BRS" : "");
+
+	k_mutex_lock(&data->tx_mutex, K_FOREVER);
+	config->irq_disable_func();
+	/* Send MB Interrupt */
+	status = Canexcel_Ip_SendFDMsg(config->instance, mb_indx, &data->tx_cbs[alloc].tx_info,
+				frame->id, (uint8_t *)&frame->data, &data->tx_msg[alloc]);
+	config->irq_enable_func();
```
Is disabling/re-enabling the IRQ needed for CANXL or was this just copy-pasted here?

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+	return 0;
+}
+
+static void nxp_s32_zcan_timming_to_canxl_timming(const struct can_timing *timing,
```
```suggestion
static void nxp_s32_zcan_timing_to_canxl_timing(const struct can_timing *timing,
```

congnguyenhuu replied:
I updated

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		return -EBUSY;
+	}
+
+	LOG_DBG("Setting CAN bitrate %d:", config->bitrate);
```
This debug log looks wrong. It uses the bitrate species in DTS instead of the timing parameters supplied to the function.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		return -EBUSY;
+	}
+
+	LOG_DBG("Setting CAN-FD bitrate %d:", config->bitrate_data);
```
Same here.

congnguyenhuu replied:
I removed it

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+	int alloc;
+	void *arg;
+
+	switch (eventType) {
+	case CANEXCEL_EVENT_TX_WARNING:
+		LOG_DBG("Tx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_RX_WARNING:
+		LOG_DBG("Rx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_BUSOFF:
+		LOG_DBG("Bus Off (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR:
+		LOG_DBG("Error Format Frames (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR_FD:
+		LOG_DBG("Error Data Phase (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_PASSIVE:
+		LOG_DBG("Error Passive (error 0x%x)", u32SysStatus);
+		break;
```
Could this be used for implementing CAN stats support?

congnguyenhuu replied:
I added support stats for them

congnguyenhuu replied:
I left out the support https://github.com/zephyrproject-rtos/zephyr/pull/52450#discussion_r1178313786
so can't use the current implementing CAN stats support

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		if (atomic_test_bit(data->rx_allocs, alloc)) {
+			nxp_s32_msg_data_to_zcan_frame(data->rx_msg[alloc], &frame);
+
+			LOG_DBG("%s: Received %d bytes Rx Mb %d, "
+				"Rx Id: 0x%x, "
+				"Id type: %s %s %s %s"
+				, dev->name, can_dlc_to_bytes(frame.dlc)
+				, buffidx, frame.id
+				, !!(frame.flags & CAN_FRAME_IDE) ?
+								"extended" : "standard"
+				, !!(frame.flags & CAN_FRAME_RTR) ? "RTR" : ""
+				, !!(frame.flags & CAN_FRAME_FDF) ? "FD frame" : ""
+				, !!(frame.flags & CAN_FRAME_BRS) ? "BRS" : "");
```
Please fix formatting here as well.

congnguyenhuu replied:
I updated format

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+		tx_func = data->tx_cbs[alloc].function;
+		LOG_DBG("%s: Sent Tx Mb %d", dev->name, buffidx);
+		if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+			tx_func(dev, 0, data->tx_cbs[alloc].arg);
```
It's good practice to check for `tx_func != NULL` prior to calling it.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+				, !!(frame.flags & CAN_FRAME_FDF) ? "FD frame" : ""
+				, !!(frame.flags & CAN_FRAME_BRS) ? "BRS" : "");
+
+			rx_func(dev, &frame, data->rx_cbs[alloc].arg);
```
Same here.

congnguyenhuu replied:
I updated

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+			status = Canexcel_Ip_ReceiveFD(config->instance, buffidx,
+							&data->rx_msg[alloc], FALSE);
+			if (status != CANEXCEL_STATUS_SUCCESS) {
+				LOG_ERR("Can't enable interrupt MB %d for next message!", buffidx);
```
How should this error be handled?

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+			&can_nxp_s32_driver_api);
+
+#if DT_NODE_HAS_STATUS(CAN_NXP_S32_NODE(0), okay)
+CAN_NXP_S32_INIT_DEVICE(0)
```
Nit: Using instances like this is HAL specific and does not follow the Zephyr design principles. You should at least check for the right compatible string and not just rely on the nodelabel.

manuargue replied:
I agree. We are planning to address this issue for all S32 peripheral drivers added so far. The limitation is that the HAL is designed in such a way that needs to know the peripheral hw instance index, sometimes at build time. In previous reviews we had the blessing to merge as it is using nodelabel, so I would suggest keeping this topic to be addressed as an improvement for the whole S32 port in a separate PR, if possible.

henrikbrixandersen replied:
Sounds good. Not a blocker from my side, just commenting on it for the record.

manuargue commented on dts/bindings/can/nxp,s32-canxl.yaml:
```diff
@@ -0,0 +1,27 @@
+
+description: NXP S32 CANXL controller
+
+compatible: "nxp,s32-canxl"
```
As of my knowledge, there’s no other device ported to Zephyr that has this peripheral, except from the S32 family. On the other hand, it wouldn’t be the only binding that have a family prefix but it is also used in SoCs of another families (I’ve seen this for some kinetis and imx peripherals). I agree that if we can avoid that from the beginning it would be cleaner. But so far we don’t have this information.

dleach02 commented:
@henrikbrixandersen, thank you for your feedback and review on this PR. Can you highlight which items are still considered a blocker so the author can address them?

manuargue commented on boards/arm/s32z270dc2_r52/doc/index.rst:
```diff
@@ -128,6 +130,13 @@ single Virtual SI (VSI). The rest of the VSI's shall be assigned to different
+Controller Area Network (CAN)
+=============================
+
+Currently, CAN driver supports classic (CAN 2.0) and CAN FD mode,
```
Nit: remove trailing comma and rephrase `..mode. Remote transmission…`

manuargue reviewed (changes requested):
Just a minor request in the docs

congnguyenhuu commented on boards/arm/s32z270dc2_r52/doc/index.rst:
```diff
@@ -128,6 +130,13 @@ single Virtual SI (VSI). The rest of the VSI's shall be assigned to different
 cores of the system. Refer to :ref:`nxp_s32_netc-samples` to learn how to
 configure the Ethernet network controller.
 
+Controller Area Network (CAN)
```
I added a note for that

congnguyenhuu commented on drivers/can/can_nxp_s32_canxl.h:
```diff
@@ -0,0 +1,93 @@
+ */
+
+#ifndef ZEPHYR_DRIVERS_CAN_NXP_S32_CANXL_H_
+#define ZEPHYR_DRIVERS_CAN_NXP_S32_CANXL_H_
```
I updated

congnguyenhuu commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1020 @@
+{
+	const struct can_nxp_s32_config *config = dev->config;
+
+	__ASSERT(rate != NULL, "Rateptr to get core clock is NULL");
```
I updated this

manuargue reviewed (dismissed):
Hi @henrikbrixandersen, thanks for the thoroughly review. Could you please have another look to the changes to see if there are still unresolved comments? We would like to have this feature included in v3.4 if possible. We can also consider having a first working version merged, and address further improvements with a follow up PR.

dleach02 commented:
@henrikbrixandersen can you return to this PR to determine if your change requests have been addressed?

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1086 @@
+		return -ENOTSUP;
+	}
+#else
+	if ((frame->flags & ~CAN_FRAME_IDE) != 0) {
```
Is this correct?

henrikbrixandersen commented on include/zephyr/drivers/can.h:
```diff
@@ -473,6 +477,12 @@ STATS_SECT_ENTRY32(crc_error)
 STATS_SECT_ENTRY32(form_error)
 STATS_SECT_ENTRY32(ack_error)
 STATS_SECT_ENTRY32(rx_overrun)
+STATS_SECT_ENTRY32(tx_warning)
+STATS_SECT_ENTRY32(rx_warning)
+STATS_SECT_ENTRY32(busoff)
+STATS_SECT_ENTRY32(frame_error)
+STATS_SECT_ENTRY32(fd_error)
+STATS_SECT_ENTRY32(passive_error)
```
Please leave these additions out.

congnguyenhuu replied:
I left out the update for support counter statistic

congnguyenhuu commented:
@henrikbrixandersen Thank for revisiting this PR. I addressed all comment.

congnguyenhuu commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1097 @@
+	Canexcel_Ip_ConfigType can_nxp_s32_default_config##n = {			\
+		.rx_mbdesc = (uint8)CONFIG_CAN_NXP_S32_MAX_RX,				\
+		.tx_mbdesc = (uint8)CONFIG_CAN_NXP_S32_MAX_TX,				\
+		.CanxlMode = CANEXCEL_LISTEN_ONLY_MODE,					\
```
done https://github.com/zephyrproject-rtos/zephyr/pull/52450#discussion_r1178317471
Please point me to where this is done?

congnguyenhuu commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -0,0 +1,1097 @@
+				CANEXCEL_FD_FRAME : CANEXCEL_CLASIC_FRAME,
+		.enable_brs = !!(frame->flags & CAN_FRAME_BRS) ? TRUE : FALSE,
+		.idType = !!(frame->flags & CAN_FRAME_IDE) ?
+				CANEXCEL_MSG_ID_EXT : CANEXCEL_MSG_ID_STD,
```
I removed the setting ` .retransmission = 0 ` in `can_send()`https://github.com/zephyrproject-rtos/zephyr/pull/52450#discussion_r1178311886

congnguyenhuu commented on include/zephyr/drivers/can.h:
```diff
@@ -172,8 +172,12 @@ struct can_frame {
+	 * full width of the timestamp.
 	 */
-	uint16_t timestamp;
+	uint32_t timestamp;
```
we can support 32bits timestamp? you suggested before https://github.com/zephyrproject-rtos/zephyr/pull/52450#discussion_r1178306712

henrikbrixandersen reviewed (changes requested):
Please leave out the timestamp changes, at least for now, and I think we're good to go.

congnguyenhuu commented:
@henrikbrixandersen I left out the timestamp change. Do you have any blocker for PR?

henrikbrixandersen reviewed (approved):
Thank you, looks good - and thank you for your patience with the review process!

dleach02 commented:
@congnguyenhuu please remove the west.yml from your PR. The HAL change was merged a while ago and is part of the current Zephyr west.yml.

congnguyenhuu commented:
@dleach02 I removed it as soon as the HAL change was merged.
